composer/utils/object_store/gs_object_store.py

```diff
@@ -0,0 +1,165 @@
+        elif isinstance(e, GatewayTimeout):
+            raise ValueError(f'Time out when uploading/downloading {uri} using google cloud storage') from e
+
+    except ImportError as e:
+        raise MissingConditionalImportError(conda_package='google-cloud-storage',
+                                            extra_deps_group='google-cloud-storage',
+                                            conda_channel='conda-forge') from e
+    # Otherwise just raise the original error.
+    raise e
+
+
+class GsObjectStore(ObjectStore):
+    """Utility for uploading to and downloading from an Google Cloud bucket using :mod:`google cloud storage sdk` with either HMAC and service account authentications. If both authentiations are available, use service account authentication.
+
+    .. warning::
+
+        For best security practices, it is recommended to set credentials via environment variables or config files.
+
+        See :ref:`guide to credentials <boto3:guide_credentials>` for more information.
+
+    Args:
+        remote_dir (str, optional): Required. The URL to an Gooogle Cloud Storage object, formatted as scheme://bucket/path
+    """
+
+    def __init__(
+            self,
+            gs_root_dir: str  #  = 'gs://mosaicml-composer-tests/streaming/',
+    ) -> None:
+        try:
+            import boto3
+            from google.cloud.storage import Client
+        except ImportError as e:
+            raise MissingConditionalImportError('streaming', 'boto3') from e
+
+        if 'GOOGLE_APPLICATION_CREDENTIALS' in os.environ:
+            service_account_path = os.environ['GOOGLE_APPLICATION_CREDENTIALS']
+            self.client = Client.from_service_account_json(service_account_path)
+        elif 'GCS_KEY' in os.environ and 'GCS_SECRET' in os.environ:
+            # Create a session and use it to make our client. Unlike Resources and Sessions,
+            # clients are generally thread-safe.
+
+            session = boto3.session.Session()
+            self.client = session.client('s3',
+                                         region_name='auto',
+                                         endpoint_url='https://storage.googleapis.com',
+                                         aws_access_key_id=os.environ['GCS_KEY'],
+                                         aws_secret_access_key=os.environ['GCS_SECRET'])
+        else:
+            raise ValueError(f'GOOGLE_APPLICATION_CREDENTIALS needs to be set for ' +
+                             f'service level accounts or GCS_KEY and GCS_SECRET env variables must be set.')
+
+        from composer.utils import parse_uri
+        backend, self.bucket_name, self.prefix = parse_uri(gs_root_dir)
+        if backend == '':
+            raise ValueError("remote_dir doesn't have a valid format")
+        self.prefix = self.prefix.lstrip('/')
+
+        try:
+            self.bucket = self.client.get_bucket(self.bucket_name, timeout=10.0)  # Bucket(self.client, obj.netloc)
```

Comment:
how'd you pick this timeout value?

Reply:
Good catch. I don't have a grounding for it. Default is 60. We can use that. Wdyt?

Reply:
sounds good yeah!